___
# Atividade: Modelos probabilísticos discretos e Dados
___

## Aula 10

**Referência de Leitura:**
1. Magalhães e Lima (7ª. Edição): Seções 3.2 e 3.3

**Hoje:**
1.	Descrever modelos discretos quanto aos resultados teóricos.
2.	Explicar a utilização de modelos probabilísticos no contexto da literatura estatística.
3.	Contrastar resultados teóricos e empíricos.
4.	Fechamento do conteúdo.

**Próxima aula:**
1. Leitura prévia necessária: Magalhães e Lima (7ª. Edição): Seção 5.1 (pág. 137) e Seção 5.2 (pág. 146).
___

In [46]:
%matplotlib inline
import pandas as pd
import matplotlib.pyplot as plt

from scipy import stats #importa apenas as funções de estatísticas da biblioteca SciPy.
import itertools #importante para gerar todos os resultados possíveis!!


___
## <font color='blue'>Exercício 1</font>

Um resistor de 100K Ohms comprado na Santa Efigênia tem probabilidade de falha de 5%, segundo um fabricante ruim.

Ainda, assuma os resistores falham de forma independente entre si.

Um aluno de engenharia compra um pacote contendo $n$ resistores.

1. Assuma $n=3$ e faça os itens a seguir.
  1. Monte o espaço amostral que combine todos os resultados possíveis de Falha e não Falha desses $n$ resistores. Ainda, calcule a probabilidade de cada resultado do espaço amostral acontecer.
  2. Se $Y$ representa o número de resistores que falham em um pacotinho com $n$ resistores, então construa a distribuição de probabilidades de $Y$.
  3. Calcule a média e o desvio padrão de $Y$. Interprete os resultados.
    
2. Assuma $n=7$ e repita os itens A) a C) anteriores. No caso de $n=7$ ou maiores quantidades de resistores em um pacotinho, qual a viabilidade de construir a distribuição de probabilidades do número de resistores com defeitos quando $n$ não é pequeno?
    
3. Encontre a distribuição de probabilidades de cada possível valor de $Y$ utilizando a distribuição Binomial para $n=3$ e $n=7$. Calcule também $E(X)$ e $Var(X)$. Compare com as distribuições de probabilidades obtidas nos itens anteriores e discuta porque esses valores são iguais ou diferentes!
    
> **DISTRIBUIÇÃO BINOMIAL - TEORIA:**

> É responsável por modelar uma variável $Y$ definida pelo Número de sucessos em $n$ tentativas, cuja probabilidade de sucesso é dada por $p$. Assim, dizemos que $Y$~$Bin(n,p)$ e lê-se: $Y$ segue uma distribuição binomial com $n$ tentativas e $p$ como probabilidade de sucesso. 

> A função de probabilidade nesse caso é dada por:
$P(Y=y)=\left(
\begin{array}{c}
    n \\
    y
\end{array}\right) p^y (1-p)^{(n-y)}$

> Para um experimento ter as probabilidade dos números de sucessos modeladas por uma distribuição Binomial, o mesmo precisa ter as seguintes propriedades:
* é uma contagem de $n$ repetições (ou tentativas ou ensaios) idênticas;
* cada repetição tem apenas 2 resultados possíveis: um é denominado sucesso e o outro, fracasso;
* a probabilidade de sucesso para cada ensaio é denominada $p$ e será constante em cada repetição. Então, a probabilidade de fracasso $(1-p)$ também não varia de tentativa para tentativa;
* As tentativas são independentes.


> **DISTRIBUIÇÃO BINOMIAL - PYTHON:** <sup> [https://en.wikipedia.org/wiki/SciPy]

> Algumas funções úteis para desenvolver seu código: <sup> [http://docs.scipy.org/doc/scipy-0.16.1/reference/generated/scipy.stats.binom.html]
* $P(Y=y)$: `stats.binom.pmf(y, n, p)`
* $P(Y\leq y)$: `stats.binom.cdf(y, n, p)`
* $E(X)$ e $Var(X)$: `stats.binom.stats(n, p, moments='mv')`


In [47]:
#item 1
n=3
#item 2
#n=7

In [48]:
import itertools

#Ex A
possibilidades = ["F", "N"]
probabilidades = {"F": 0.05, "N":0.95}
arranjos = itertools.product(possibilidades, repeat=n)
espaco_amostral = list(arranjos)
probabilidades_linhas = []

for res in espaco_amostral:
    p = 1.0
    for e in res:
        p*=probabilidades[e]
    probabilidades_linhas.append(p)

print(espaco_amostral)
print(probabilidades_linhas)

[('F', 'F', 'F'), ('F', 'F', 'N'), ('F', 'N', 'F'), ('F', 'N', 'N'), ('N', 'F', 'F'), ('N', 'F', 'N'), ('N', 'N', 'F'), ('N', 'N', 'N')]
[0.00012500000000000003, 0.0023750000000000004, 0.0023750000000000004, 0.045125, 0.0023750000000000004, 0.045125, 0.045125, 0.8573749999999999]


In [49]:
#Ex B
valores_X = {}
for i in range(len(espaco_amostral)):
    falha = espaco_amostral[i].count("F")
    # Checar antes se a chave está no dicionário
    if falha in valores_X:
        valores_X[falha] += probabilidades_linhas[i]
    else:
        valores_X[falha] = probabilidades_linhas[i]

valores_X

{0: 0.8573749999999999,
 1: 0.135375,
 2: 0.007125000000000001,
 3: 0.00012500000000000003}

___
## <font color='blue'>Exercício 2</font>

Uma fábrica produz determinado veículo. O gerente de vendas fechou um negócio que implica na produção de pelo menos 110 unidades nos próximos dois meses. Para isso, será construída uma nova linha de montagem com capacidade para produzir em média 50 veículos por mês (considere meses de 30 dias). A linha funcionará ininterruptamente. Além disso, apenas um veículo é produzido por vez. Considere que um modelo Poisson seja adequado para a quantidade de veículos produzidos.
   1. Qual a probabilidade  da produção ser igual a 100 veículos nos próximos dois meses?
   2. Qual a probabilidade  da produção ser de pelo menos 100 veículos nos próximos dois meses?
   3. Qual a probabilidade da produção ser entre 100 e 110 veículos nos próximos dois meses?
   4. Se forem produzidos de pelo menos 100 veículos no próximo mês, qual a probabilidade do gerente cumprir a meta negociada?
   5. Caso o prazo seja cumprido, o lucro será da ordem de $250$ mil, caso não seja, o lucro cai para $200$ mil. Qual é o lucro esperado e seu desvio-padrão?

**Observação:** Explore os comandos de `stats.poisson`. <sup>[http://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.poisson.html]


___
## <font color='blue'>Exercício 3</font> 
### <font color='red'>**Check para dia 13/09 em DUPLA **</font>

Esse exercício irá explorar uma modelagem de dados reais.

> **Confira alguns itens obrigatórios verificados durante a vistoria do Detran**

> *Todos os veículos, novos ou velhos, precisam passar por uma vistoria todos os anos. (...) O motorista precisa estar atento a alguns itens obrigatórios. Tudo deve funcionar perfeitamente, apresentar bom estado de conservação e estar dentro do prazo de validade.*

> <sup>Fonte: http://extra.globo.com/noticias/brasil/transito-seguro/confira-alguns-itens-obrigatorios-verificados-durante-vistoria-do-detran-10190355.html</sup>

Essa matéria lista 14 itens que são inspecionados pelo Detran, dentre os quais têm-se: extintor de incêndio deve estar dentro do prazo de validade; pneus devem estar em bom estado de conservação; buzina deve funcionar perfeitamente; e cintos de segurança para todos os ocupantes do carro. Se, no final das contas, todos os 14 itens estiverem funcionando perfeitamente, o motorista irá feliz para casa assegurado de que seu carro está sem problemas com a vistoria.

1. A base de dados `Carros.txt` apresenta, para cada um dos três mil carros de passeio vistoriados, duas informações: tipo de carro (1: Popular e 2: Não Popular) e quantidade de itens vistoriados em não conformidade (que pode variar de 0 a 14). 
    1. Considerando todos os carros, analise a quantidade de itens vistoriados em não conformidade. Pela tabela de frequências relativas, calcule a média e variância dessa quantidade. Interprete-os.
    2. Assuma que a variável *Quantidade de itens em não conformidade* possa ser ajustada pelo modelo Binomial com parâmetros $n=14$ e $p=0,10$. Interprete esses parâmetros para o problema em questão e, ainda, discuta se as propriedades da distribuição Binomial estão satisfeitas para o problema aqui me questão.
    
    **Independente da sua resposta anterior, considere que a distribuição binomial seja adequada para modelar a variável de interesse nos próximos itens!**
    3. Utilizando o método `stats.binom.pmf`, obtenha a probabilidade de cada uma das quantidades de itens em não conformidade quando $n=14$ e $p=0,10$. Calcule o valor esperado e a variância da quantidade de itens em não conformidade utilizando o método `stats.binom.stats` e via resultado da Tabela 4.4 do Magalhães e Lima (pág. 125).
    4. Compare as probabilidades teóricas com as frequências relativas observadas nos dados. Para isso, tem-se, por exemplo, as duas opções abaixo para comparar se os resultados empíricos se encaixam com o modelo teórico proposto. 
        - A primeira opção gráfica contrasta a frequência relativa e a probabilidade teórica para uma determinada quantidade de itens em não conformidade! Vide Figura 3.4 do livro Magalhães e Lima (pág. 87). Ou seja, essa opção contrasta a frequência relativa em um valor de $y$ e a probabilidade pontual $P(Y=y)$.
        - A segunda contrasta a frequência relativa acumulada com a probabilidade acumulada até determinada quantidade de itens em não conformidade. Ou seja, compara a frequência relativa acumulada até $y$ com a probabilidade acumulada $P(Y\leq y)$.
    Construa os gráficos de acordo com as duas opções e refine sua conclusão quanto ao uso do modelo teórico para ajustar a variável em questão!
    5. Na prática, qual a necessidade de um modelo probabilístico já que se têm os dados observados?
    6. Considerando apenas os carros populares, o modelo Binomial com parâmetros $n=14$ e $p=0,10$ é adequado para ajustar a variável Quantidade de itens em não conformidade? Se sim, justifique; caso não, sugira novos valores para os parâmetros da distribuição ajuste aos dados.

In [50]:
import os
print('Esperamos trabalhar no diretório')
print(os.getcwd())
carros = pd.read_table('Carros.txt', sep=' ')

Esperamos trabalhar no diretório
C:\Users\Isabella\Documents\GitHub\CD\aula10


In [77]:
# A frequência relativa
frequenciar = carros.Quantidade.value_counts(True) * 100
frequenciar

1    31.666667
0    28.800000
2    22.133333
3    10.833333
4     4.633333
5     1.566667
6     0.266667
7     0.066667
8     0.033333
Name: Quantidade, dtype: float64

In [52]:
carros.Quantidade.describe()

count    3000.000000
mean        1.371333
std         1.256969
min         0.000000
25%         0.000000
50%         1.000000
75%         2.000000
max         8.000000
Name: Quantidade, dtype: float64

In [53]:
# A média
tamanho = carros.shape
count = tamanho[0]
media = sum(carros.Quantidade)/count
media

1.3713333333333333

In [54]:
# A variância
variancia = carros.Quantidade.var()
variancia

1.5799715460709343

Com os dados acima, podemos perceber que a média é de 1.37 itens vistoriados em não conformidade por carro que passou pela vistoria. Um número não muito alto, pouco mais de 1 item por carro não está conforme o previsto nas novas regras do Detran.
A variância nos mostra o quanto os dados "números de itens vistoriados em não conformidade" estão afastados de seu valor esperado/esperança, ou seja, o "quão longe" os números estão da média. O resultado obtido foi 1.57, um valor não muito alto, indicando que eles estão mais do que 1 item de distância do esperado, porém menos do que 2.

In [61]:
#B
p = 0.1 #probabilidade de dar não conformidade
n = 14 #tentativas/ itens vistoriados
import itertools

#Ex A
possibilidades = ["C", "N"]
probabilidades = {"C": 0.9, "N":0.1}
arranjos = itertools.product(possibilidades, repeat=n)
espaco_amostral = list(arranjos)
probabilidades_linhas = []

for res in espaco_amostral:
    j = 1.0
    for e in res:
        j*=probabilidades[e]
    probabilidades_linhas.append(j)
    
valores = {} #acumular quantidade
for i in range(len(espaco_amostral)):
    quantidade = espaco_amostral[i].count("N")
    # Checar antes se a chave está no dicionário
    if quantidade in valores:
        valores[quantidade] += probabilidades_linhas[i]
    else:
        valores[quantidade] = probabilidades_linhas[i]

In [74]:
print("Probabilidade por método tradicional")
for k in valores:
    print("Sucessos: ", k, " Probabilidade", "%6.5f "%valores[k])
    
print()

print("Probabilidade pelo método do binômio")
for y in valores:
    binomio = stats.binom.pmf(y, n, p)
    print("Sucessos: ", y, " Probabilidade", "%6.5f "%binomio)


Probabilidade por método tradicional
Sucessos:  0  Probabilidade 0.22877 
Sucessos:  1  Probabilidade 0.35586 
Sucessos:  2  Probabilidade 0.25701 
Sucessos:  3  Probabilidade 0.11423 
Sucessos:  4  Probabilidade 0.03490 
Sucessos:  5  Probabilidade 0.00776 
Sucessos:  6  Probabilidade 0.00129 
Sucessos:  7  Probabilidade 0.00016 
Sucessos:  8  Probabilidade 0.00002 
Sucessos:  9  Probabilidade 0.00000 
Sucessos:  10  Probabilidade 0.00000 
Sucessos:  11  Probabilidade 0.00000 
Sucessos:  12  Probabilidade 0.00000 
Sucessos:  13  Probabilidade 0.00000 
Sucessos:  14  Probabilidade 0.00000 

Probabilidade pelo método do binômio
Sucessos:  0  Probabilidade 0.22877 
Sucessos:  1  Probabilidade 0.35586 
Sucessos:  2  Probabilidade 0.25701 
Sucessos:  3  Probabilidade 0.11423 
Sucessos:  4  Probabilidade 0.03490 
Sucessos:  5  Probabilidade 0.00776 
Sucessos:  6  Probabilidade 0.00129 
Sucessos:  7  Probabilidade 0.00016 
Sucessos:  8  Probabilidade 0.00002 
Sucessos:  9  Probabilidade 0.00

Os resultados obtidos calculando normalmente a probabilidade para nossa situação em questão e calculando com o binômio são semelhantes, o que comprova que nesse caso, as propriedades da distribuição Binomial  são satisfeitas.

In [85]:
#C Valor esperado e variância segundo distribuição binomial
stats.binom.stats(n, p, moments='mv')

(array(1.4000000000000001), array(1.2600000000000002))

In [86]:
#C Segundo a Tabela 4.4 do Magalhães e Lima
Valor_esperado = n*p
variancia = n*p*(1-p)
print("Valor esperado: ", Valor_esperado)
print("Variância: ", variancia)

Valor esperado:  1.4000000000000001
Variância:  1.2600000000000002


Os valores obtidos com a função de distribuição Binomial e com a Tabela 4.4 do Magalhães e Lima são idênticas

In [83]:
#D Comparação entre a frequência relativa de ocorrência de não conformidade com a probabilidade da ocorrência
print(frequenciar)
print()
print("Probabilidade pelo método do binômio")
for y in range(0,9):
    binomio = stats.binom.pmf(y, n, p)
    binomioo = binomio*100
    print(y, " Probabilidade", "%6.5f "%binomioo)

1    31.666667
0    28.800000
2    22.133333
3    10.833333
4     4.633333
5     1.566667
6     0.266667
7     0.066667
8     0.033333
Name: Quantidade, dtype: float64

Probabilidade pelo método do binômio
0  Probabilidade 22.87679 
1  Probabilidade 35.58612 
2  Probabilidade 25.70109 
3  Probabilidade 11.42271 
4  Probabilidade 3.49027 
5  Probabilidade 0.77562 
6  Probabilidade 0.12927 
7  Probabilidade 0.01642 
8  Probabilidade 0.00160 


Existe uma proximidade entre os valores esperados obtidos, apesar deles não serem exatamente como os dados da frequência relativa.

In [95]:
# D comparação frequencia relativa acumulada e probabilidade acumulada
print("Frequencia relativa acumulada")
fa = 0
for i in range(0,9):
    freq = carros.Quantidade.value_counts(True)[i]
    print("%6.5f "%freq)
    fa += freq
print(fa)

print()
print("Probabilidade acumulada")
pa = 0
for y in range(0,9):
    prob = binomio = stats.binom.pmf(y, n, p)
    print("%6.5f "%binomio)
    pa += binomio
    
print(pa)

Frequencia relativa acumulada
0.28800 
0.31667 
0.22133 
0.10833 
0.04633 
0.01567 
0.00267 
0.00067 
0.00033 
1.0

Probabilidade acumulada
0.22877 
0.35586 
0.25701 
0.11423 
0.03490 
0.00776 
0.00129 
0.00016 
0.00002 
0.999998749435


A comparação nesse tópico também não é exata porém os valores de frequência relativa acumulada e probabilidade acumulada são próximos.

E) A necessidade de um modelo probabilístico é 